presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+public class AggregationStatsRule
+        extends SimpleStatsRule
+{
+    private static final Pattern<AggregationNode> PATTERN = Pattern.typeOf(AggregationNode.class);
+
+    public AggregationStatsRule(StatsNormalizer normalizer)
+    {
+        super(normalizer);
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    protected Optional<PlanNodeStatsEstimate> doCalculate(PlanNode node, StatsProvider statsProvider, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        AggregationNode aggregationNode = (AggregationNode) node;
+
+        if (aggregationNode.getGroupingSets().size() != 1) {
+            return Optional.empty();
+        }
+
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(aggregationNode.getSource());
+
+        List<Symbol> groupBySymbols = getOnlyElement(aggregationNode.getGroupingSets());
+
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+        for (Symbol groupBySymbol : groupBySymbols) {
+            SymbolStatsEstimate symbolStatistics = sourceStats.getSymbolStatistics(groupBySymbol);
+            result.addSymbolStatistics(groupBySymbol, symbolStatistics.mapNullsFraction(nullsFraction -> {
+                if (isPositiveOrNan(nullsFraction)) {
+                    double distinctValuesCount = symbolStatistics.getDistinctValuesCount();
+                    return 1.0 / (distinctValuesCount + 1);
```

Comment:
>  or `(1.0 / (distinctValuesCount + 1)) * aggregtaion_output_estimated_rows_count` might be better.

I guess your formula has a typo -- since we calculate nulls _fraction_, it should go without `* aggregtaion_output_estimated_rows_count` part, right? But then, isn't it what's written here? Please clarify.

Reply:
Let me put here a more detailed explanation.

When doing a group by by a single column, `null` values are folded into a single `null`. So the `1 / (ndv + 1)` is perfect.

However when grouping by multiple columns, values are folded on row basis.

```
SELECT *
FROM (values (null, 1), (null, 2), (null, 3), (null, 4), ... ,(null, N)) as t(x,y)
GROUP BY (x, y)
```
results in 
```
null | 1
null | 2
null | 3
null | 4
...
null | N
```

In this case a `null_fraction * output_rows` is a better estimate than a `1 / (ndv + 1)`. Not ideal though, as `null` might correlate with other columns. But because we don't know anything about correlation, the `output_rows` estimate is also off.

Reply:
UPD:

Based on the offline discussion. What we are doing right now seems to be the best.